docs/cpan-dependency-confusion.md

```diff
@@ -0,0 +1,238 @@
+
+## Background
+"Dependency confusion" takes advantage of mixing (semi-)private modules with public CPAN modules.
+
+In Perl ecosystem, it could be one or more of the following options:
+- Using a "DarkPAN" along with CPAN
+- Using a "GreyPAN" along with CPAN
+- Mixing manual installs (`make install`, "blib install"...) along with CPAN
+
+See [About the various PANs](https://www.olafalders.com/2019/02/19/about-the-various-pans/) for details about DarkPAN and GreyPAN.
+
+An attacker can "claim" a public namespace, install a high version (to take precedence in index over a smaller local version) and wait for vulnerable setups to install (and execute) their newly uploaded module.
+
+## General about CPAN dependency management
+### The mirror versus the indexer
+The notions of "mirror" and "indexer":
+- "mirror" - where module tarballs are stored
+- "indexer" - translate module to path and version
+
+These notions are not always well differenciated and their meaning heavily depends the CPAN client. For instance, from a pool (urllist), `CPAN.pm` will use only one mirror for both index and download of modules. On the other hand, `cpanm` will use MetaDB as index in priority and mirrors to download, and these notions are well decoupled. 
+
+Not all indexers are equals: 02packages only contains latest version of each module, making it [difficult to specify more complex constraints](https://github.com/miyagawa/Carmel/issues/55) (pinned or ranges).
+
+### General characteristics about CPAN/PAUSE and installers
+- Module installs are ([often](https://blogs.perl.org/users/shoichi_kaji1/2017/03/make-your-cpan-module-static-installable.html)) "dynamic" (= some dependencies are only known at install)
+- Claiming namespace is case insensitive but index/install is case sensitive (important for TypoSquatting or confusions attacks based on modules/filenames collisions)
+
+### Installer are not zealous
+#### "leafs" vs "requirements"
+CPAN installers treats differently a "leaf" versus a "requirement".
+
+Specifying a module as a "leaf", e.g. `cpanm Requirement` is different from having it specified as a dependency (Makefile.PL/Build.PL, META.*, cpanfile...)
+
+- When explicitely targetting to install `Requirement`, `cpanm` (and others) will install latest, upgrading the module if present
+- When `Requirement` module is specified as a dependency, `cpanm` will not upgrade module if installed version satisfies the constraints
+
+Installing with `cpanm --installdeps .` will not install anything:
+1. Requirement@0.01
+2. cpanm --installdeps .
+3. Requirement@0.01
+
+But `cpanm Requirement` will update `Requirement`:
+1. Requirement@0.01
+2. cpanm Requirement
+3. Requirement@0.02
+
+Note: This is not specific to CPAN ecosystem, for instance `pip` and `gem` are behaving the same.
+
+#### `--skip-satisfied`
+But `cpanm Requirement` will update `Requirement`:
+1. Requirement@0.01
+2. cpanm --skip-satisfied Requirement
+3. Requirement@0.01
```

Comment:
Shouldn‘t this be 0.02?

Reply:
No, it's changing the behavior to do *not* upgrade:

```
$ cpanm -l local --skip-satisfied Requirement
You have Requirement (0.01)
```